### ISSUE TITLE
Progress updates for execution 

### ISSUE THREAD

**Issue by quintona** [nteract/hydrogen#314]
At present, any execution simply displays a simple progress update during the execution. This works really well for short running executions, for calls that take more time (complex executions, that launch cluster wide or involve some lengthy process, like training an ML model) there is no way of differentiating between a failure (it will never return) from something that is just running for a long time. 

I have looked at the iPython kernel docs and I can't see anything that would allow the execution to return some progress (maybe something in the Kernel event loop). 

It would be really good if some kind of progress could be displayed for the longer running executions?

**Comment by lgeiger** [comment]
I don't think this is possible to do in an language independent way, at least for now.

At least for python you can trie http://pythonhosted.org/progressbar2/ or https://github.com/tqdm/tqdm.

**Comment by n-riesco** [comment]
The way to determine the kernel didn't stall is checking the kernel heartbeat.

Currently, Hydrogen doesn't check the kernel hearbeat. We could implement this, but unfortunately the kernel heartbeat has been deprecated. See https://github.com/jupyter/enhancement-proposals/pull/14 .

**Comment by maximsch2** [comment]
[Juno](https://github.com/JunoLab/atom-julia-client) has progress bar support (but it is Julia-specific and depends on their library). Their UI is pretty neat and it would be great to replicate it in Hydrogen in some language-agnostic way. 

For example, tqdm and ProgressBar.jl (progress bar libraries for Python and Julia) can use the same output format that would be parsed by Hydrogen (and maybe Jupyter Notebooks as well) and rendered as a progress bar instead of the console output.

**Comment by rgbkrk** [comment]
This is on the [Spark roadmap for Jupyter](https://github.com/jupyter/roadmap/blob/master/spark.md) at least, with updatable displays being the [primary avenue to implement this](https://github.com/jupyter/roadmap/blob/master/spark.md#asynchronousbackground-outputs-).
